src/lightning/fabric/utilities/data.py

```diff
@@ -196,7 +197,8 @@ def _dataloader_init_kwargs_resolve_sampler(
                 )
 
             batch_sampler = _reinstantiate_wrapped_cls(batch_sampler, *args, **kwargs)
-        else:
+        elif isinstance(batch_sampler, BatchSampler):
+            # This is a PyTorch `BatchSampler` for which we could not capture the init args
```

Comment:
I think this is a breaking change. Doesn't this change mean that objects that implement the exact batch sampler API but don't subclass it are now unsupported?

Reply:
I believe so, and https://github.com/Lightning-AI/lightning/pull/18402/files#diff-092f8fcfded5ca85ce38376661e86413badaa7454530cc9496f7bdaaa655b9dbL216-L220 suggests this.

The best way to check would be to write a test for this and see if it fails now

Reply:
I can't determine where your link should be pointing. Could you tell me which file?

Reply:
@carmocca I was able to add that test and support the case you were requesting. In summary, we handle these cases

- the sampler is a subclass of BatchSampler: We capture the init args and re-instantiate with updated args
- the sampler is not a subclass of BatchSampler but looks like it conforms to the BatchSampler interface (having batch_size and drop_last defined): We re-instantiate based on these args only and hope for the best
- the sampler is not a subclass of BatchSampler and doesn't look like it conforms to any interface we can handle: This PR raises a new error now with the suggestion to set `use_distributed_sampler=False`

Reply:
Thanks